### PR TITLE
fix: update tracked group when going back to previously tracked group

### PR DESCRIPTION
**Pull request by pedroslopez**
Switching back to the group that was tracked when loading the page wouldn't cause any changes.

### CODE REVIEW

**Comment by pedroslopez** [comment]
This works, though I'm not sure if this is something that should be decided in the UI (whether to track/untrack a group)

File: ui/ui-enterprise/pages/destination/[id]/data.tsx [hunk 1]
```diff
@@ -77,7 +80,7 @@ export default function Page(props) {
```
Comment (evantahler) [inline ui/ui-enterprise/pages/destination/[id]/data.tsx:79]: This is a lot of logic to have in the UI.  It's my fault for setting it up this way.

I think the path forward is to have one API action (eg: `destination:edit`, `PUT /destination/id`) that can also accept a `trackedGroupId`, and we handle this logic in the Action.  This way the form could just submit whatever it has.

* if `trackedGroupId === destination.groupId` do nothing
* if `trackedGroupId === null || ""` call `desintaion.untrack()`
* if `trackedGroupId !== destination.groupId` call `destinatio.track(trackedGroupId)`

**Review by evantahler — CHANGES_REQUESTED** [review]
I'm glad you found what the problem was... and I think you are right... the UI is probably overloaded at this point 😦

File: ui/ui-enterprise/pages/destination/[id]/data.tsx (hunk 1 above)
Comment (evantahler) [reply ui/ui-enterprise/pages/destination/[id]/data.tsx:79]: This may also mean we can remove the Actions for destination:track and destination:untrack

Comment (evantahler) [reply ui/ui-enterprise/pages/destination/[id]/data.tsx:79]: I think it's also confusing that the save button on this page /sometimes/ also means "re-export everyone" (that's the case that does `await execApi("post", "/destination/${id}/export", { force: true });`.  

That should probably be moved to it's own button called "re-export group members" or similar.

**Comment by pedroslopez** [comment]
Updated the edit action to handle tracking/untracking groups. I also made the force export a separate button, but the only thing I'm worried about here is it may give the impression that you _need_ to click this after updating the tracked group for an export to trigger

image

File: core/src/actions/destinations.ts [hunk 2]
```diff
@@ -176,6 +177,20 @@ export class DestinationEdit extends AuthenticatedAction {
 
     await destination.update(params);
 
+    if (
+      params.trackedGroupId &&
+      params.trackedGroupId !== "_none" &&
```
Comment (evantahler) [inline core/src/actions/destinations.ts:182]: This `'_none'` thing is kind of janky, but the alternative would be to have another param like `unTrackedGroupId`, and the whole point is to remove this logic from the client... so I'm 👍 with it!

File: ui/ui-components/utils/apiData.ts [hunk 3]
```diff
@@ -343,12 +341,6 @@ export namespace Actions {
   export type DestinationProfilePreview = AsyncReturnType<
     typeof DestinationProfilePreview.prototype.runWithinTransaction
   >;
-  export type DestinationTrackGroup = AsyncReturnType<
-    typeof DestinationTrackGroup.prototype.runWithinTransaction
-  >;
-  export type DestinationUnTrackGroup = AsyncReturnType<
-    typeof DestinationUnTrackGroup.prototype.runWithinTransaction
-  >;
```
Comment (evantahler) [inline ui/ui-components/utils/apiData.ts:351]: Glad you found these!  This file is how the UI's know the TS Types of the Action's responses.

File: core/__tests__/actions/destinations.ts [hunk 4]
```diff
@@ -514,7 +513,12 @@ describe("actions/destinations", () => {
           })
         );
 
-        await specHelper.runAction("destination:unTrackGroup", connection);
+        connection.params = {
+          csrfToken,
+          id,
+          trackedGroupId: "_none",
+        };
+        await specHelper.runAction("destination:edit", connection);
```
Comment (evantahler) [inline core/__tests__/actions/destinations.ts]: Might want to confirm here that the destination you get back from `runAction` (or the model directly) isn't tracking a group any more

File: core/__tests__/actions/destinations.ts [hunk 5]
```diff
@@ -472,17 +465,23 @@ describe("actions/destinations", () => {
...
       });
 
       test("an administrator can export the members of a destination with a forced group run", async () => {
-        await specHelper.runAction("destination:trackGroup", connection);
+        connection.params = {
+          csrfToken,
+          id,
+          trackedGroupId: group.id,
+        };
+        await specHelper.runAction("destination:edit", connection);
```
Comment (evantahler) [inline core/__tests__/actions/destinations.ts]: might want to confirm here that the destination now has a groupId tracked.

**Review by evantahler — APPROVED** [review]
Nice.
Screen Shot 2021-05-05 at 2 00 31 PM

I added some minor testing comments.  As for the second button, I think making it `outline` does hint that it's it's own thing. You could also put it next to vs below the main CTA which may help show that it's a choice?

File: core/src/actions/destinations.ts (hunk 2 above)
Comment (pedroslopez) [reply core/src/actions/destinations.ts:182]: Yeeah, those were my thoughts as well

**Pull request merged**